### Edgar Moises Hernandez-Gonzalez
#### 19/11/19-25/11/19
#### Clasificacion de alfa de los canales C3 y C4 de MI-EEG con SVM o Random Forest
##### 10 ejecuciones independientes para obtener la media y desviacion estandar

In [1]:
import numpy as np
import pandas as pd
#from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import cohen_kappa_score, confusion_matrix, precision_recall_fscore_support

In [2]:
x_train = pd.read_csv("MI-EEG-Alfa-A01T.csv", header=None)
x_test = pd.read_csv("MI-EEG-Alfa-A01E.csv", header=None)
y_train = pd.read_csv("EtiquetasA01T.csv", header=None)
y_test = pd.read_csv("EtiquetasA01E.csv", header=None)

In [3]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(288, 510)
(288, 510)
(288, 1)
(288, 1)


In [4]:
#escalar los datos para que tengan media = 0 y desviacion estandar = 1
scaler = StandardScaler().fit(x_train)
x_train = scaler.transform(x_train)
x_test = scaler.transform(x_test)

In [5]:
array_test_acc = []
array_kappa = []
array_precision = []
array_recall = []
array_f1 = []
for i in range(10):
    print("Iteration:", i+1);
    
    #clf = LinearSVC().fit(x_train, y_train)
    clf = RandomForestClassifier().fit(x_train, y_train)
    test_acc = clf.score(x_test, y_test)
    print('Accuracy:', test_acc)
    array_test_acc.append(test_acc)
    y_pred = clf.predict(x_test)
    kappa = cohen_kappa_score(y_test, y_pred)
    print("kappa: ", kappa)
    array_kappa.append(kappa)
    matriz_confusion = confusion_matrix(y_test, y_pred)
    print("confusion matrix:\n", matriz_confusion)
    metricas = precision_recall_fscore_support(y_test, y_pred, average='macro')
    print(metricas)
    array_precision.append(metricas[0])
    array_recall.append(metricas[1])
    array_f1.append(metricas[2])
    
print()    
print("Resultados:")
print("accuracy:", array_test_acc)
print("kappa:", array_kappa)
print("precision:", array_precision)
print("recall:", array_recall)
print("f1:", array_f1)

Iteration: 1
Accuracy: 0.2569444444444444
kappa:  0.0092592592592593
confusion matrix:
 [[16 25 18 13]
 [23 20 15 14]
 [21 17 22 12]
 [18 23 15 16]]
(0.2614042819925173, 0.2569444444444444, 0.25748451556590324, None)
Iteration: 2
Accuracy: 0.2638888888888889
kappa:  0.01851851851851849
confusion matrix:
 [[29 16 15 12]
 [24 21 15 12]
 [27 22 14  9]
 [25 23 12 12]]
(0.26223867595818817, 0.2638888888888889, 0.2560722734186717, None)
Iteration: 3
Accuracy: 0.2986111111111111
kappa:  0.06481481481481477
confusion matrix:
 [[35 15  8 14]
 [20 21 10 21]
 [26 17 13 16]
 [23 22 10 17]]
(0.2959029080675422, 0.2986111111111111, 0.28909679921848064, None)
Iteration: 4
Accuracy: 0.2708333333333333
kappa:  0.02777777777777779
confusion matrix:
 [[33 14 15 10]
 [32 16 16  8]
 [30 13 19 10]
 [22 21 19 10]]
(0.26764287390717595, 0.2708333333333333, 0.25895554869272014, None)
Iteration: 5
Accuracy: 0.22916666666666666
kappa:  -0.02777777777777768
confusion matrix:
 [[30 17 11 14]
 [27 15 18 12]
 [26 19

C:\Users\EdgarMoises\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\EdgarMoises\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove the CWD from sys.path while we load stuff.
C:\Users\EdgarMoises\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\EdgarMoises\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove t

Accuracy: 0.3368055555555556
kappa:  0.1157407407407407
confusion matrix:
 [[36 13 12 11]
 [29 23  4 16]
 [19 14 20 19]
 [27 18  9 18]]
(0.347063515721604, 0.3368055555555556, 0.33215006893888277, None)
Iteration: 8
Accuracy: 0.2847222222222222
kappa:  0.04629629629629628
confusion matrix:
 [[24 28 10 10]
 [25 28 11  8]
 [25 20 17 10]
 [25 20 14 13]]
(0.29452178918642336, 0.2847222222222222, 0.279579282920404, None)
Iteration: 9
Accuracy: 0.25
kappa:  0.0
confusion matrix:
 [[24 17 19 12]
 [32 20 14  6]
 [30 18 16  8]
 [20 18 22 12]]
(0.26038232085990387, 0.25, 0.24687075806796843, None)
Iteration: 10
Accuracy: 0.2604166666666667
kappa:  0.01388888888888884
confusion matrix:
 [[20 20 17 15]
 [22 24 12 14]
 [15 23 18 16]
 [20 22 17 13]]
(0.25869777803076427, 0.2604166666666667, 0.25782472603865003, None)

Resultados:
accuracy: [0.2569444444444444, 0.2638888888888889, 0.2986111111111111, 0.2708333333333333, 0.22916666666666666, 0.2465277777777778, 0.3368055555555556, 0.2847222222222222, 

C:\Users\EdgarMoises\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\EdgarMoises\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove the CWD from sys.path while we load stuff.
C:\Users\EdgarMoises\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\EdgarMoises\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove t

In [6]:
print("Mean: %.4f" % np.mean(array_test_acc))
print("std: (+/- %.4f)" % np.std(array_test_acc))
print("kappa: %.4f" % np.mean(array_kappa))
print("precision: %.4f" % np.mean(array_precision))
print("recall: %.4f" % np.mean(array_recall))
print("f1: %.4f" % np.mean(array_f1))

Mean: 0.2698
std: (+/- 0.0290)
kappa: 0.0264
precision: 0.2701
recall: 0.2698
f1: 0.2624


#### SVC
##### Mean: 0.3646
##### std: (+/- 0.0000)
##### kappa: 0.1528
##### precision: 0.3682
##### recall: 0.3646
##### f1: 0.3645

#### LinearSVC
##### Mean: 0.2812
##### std: (+/- 0.0000)
##### kappa: 0.0417
##### precision: 0.2871
##### recall: 0.2812
##### f1: 0.2821

#### Random Forest
##### Mean: 0.2698
##### std: (+/- 0.0290)
##### kappa: 0.0264
##### precision: 0.2701
##### recall: 0.2698
##### f1: 0.2624